fvm/evm/handler/handler.go

```diff
@@ -121,6 +121,126 @@ func (h *ContractHandler) Run(rlpEncodedTx []byte, coinbase types.Address) *type
 	return res.ResultSummary()
 }
 
+// BatchRun tries to run batch of rlp-encoded transactions and
+// collects the gas fees and pay it to the coinbase address provided.
+// All transactions provided in the batch are included in a single block,
+// except for invalid transactions
+func (h *ContractHandler) BatchRun(rlpEncodedTxs [][]byte, coinbase types.Address) []*types.ResultSummary {
+	res, err := h.batchRun(rlpEncodedTxs, coinbase)
+	panicOnError(err)
+
+	resSummary := make([]*types.ResultSummary, len(res))
+	for i, r := range res {
+		resSummary[i] = r.ResultSummary()
+	}
+	return resSummary
+}
+
+func (h *ContractHandler) batchRun(rlpEncodedTxs [][]byte, coinbase types.Address) ([]*types.Result, error) {
+	// prepare block view used to run the batch
+	ctx, err := h.getBlockContext()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx.GasFeeCollector = coinbase
+	blk, err := h.emulator.NewBlockView(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	bp, err := h.blockStore.BlockProposal()
+	if err != nil {
+		return nil, err
+	}
+
+	// decode all transactions and calculate total gas limit
+	var totalGasLimit types.GasLimit
+	batchLen := len(rlpEncodedTxs)
+	txs := make([]*gethTypes.Transaction, batchLen)
+
+	for i, rlpEncodedTx := range rlpEncodedTxs {
+		tx, err := h.decodeTransaction(rlpEncodedTx)
+		if err != nil {
+			return nil, err
+		}
+
+		txs[i] = tx
+		totalGasLimit += types.GasLimit(tx.Gas())
+	}
+
+	// check if all transactions in the batch are below gas limit
+	err = h.checkGasLimit(totalGasLimit)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := blk.BatchRunTransactions(txs)
+	if err != nil {
+		return nil, err
+	}
+
+	// saftey check for result
+	if len(res) == 0 {
+		return nil, types.ErrUnexpectedEmptyResult
+	}
+
+	bp.PopulateReceiptRoot(res)
+
+	// meter all the transaction gas usage and append hashes to the block
+	for _, r := range res {
+		// meter gas anyway (even for invalid or failed states)
+		err = h.meterGasUsage(r)
+		if err != nil {
+			return nil, err
+		}
+
+		// include it in a block only if valid (not invalid)
+		if !r.Invalid() {
+			bp.AppendTxHash(r.TxHash)
+		}
```

Comment:
I suggest we do the reverse, is successful ro failed (to not have to deal with the case that status is unknown)

Reply:
Not sure how otherwise we would do that? If a transaction has a validation error we don't otherwise if it has vm error or not we include it. There's only two options possible, if validation error not include otherwise include it. Not sure what the unknown status would be?